### PR TITLE
Allow arbitrary Lisp code in execute-command.

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This imbues our `execute-command` with the ability to execute arbitrary Lisp code, all with helpful prompt suggestions. Some part of the inspiration comes from Interlisp shell that allowed to use Lisp expressions without the outer layer of parentheses and was extrememly helpful in suggesting code pieces. While this patch is by no means a shot at the level of convenience there was, it's close enough to be usable as an inline Lisp REPL :)

Started because of #1421, but doesn't (yet?) close it.

And here are some screenshots to tease you with the suggestions it shows:
exec-var-sug
exec-fun-sug
exec-expr-parsing

@aadcg, @Ambrevar, @jmercouris, what do you guys think of it?

# Discussion

Do we need `execute-extended-command` now that we have this? I mean, `execute-extended-command` allows type-checking the code one runs, but do we need to watch user actions that closely, or should we allow them to run things and let them fail?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [X] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [x] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

EDIT: Add screenshots.

### CODE REVIEW

File: source/manual.lisp [hunk 1]
```diff
@@ -196,12 +196,6 @@ follows.")
                            (mapcar (lambda (engine) (apply 'make-search-engine engine))
                                    *my-search-engines*)))))")))
 
-      (:nsection :title "URL-dispatchers"
-        (:p "You can configure which actions to take depending on the URL to be
-loaded.  For instance, you can configure which Torrent program to start to load
-magnet links.  See the" (:nxref :function 'url-dispatching-handler) " function
-documentation."))
-
```
Comment (Ambrevar) [inline source/manual.lisp:204]: Why remove this?

File: source/command-commands.lisp [hunk 2]
```diff
@@ -53,15 +53,90 @@ from a key binding.")
...
+   (prompter:filter-preprocessor
+    (lambda (suggestions source input)
+      (unless (uiop:emptyp input)
+        (or
+         (ignore-errors
+          (let* ((proper-input (if (and (str:starts-with-p "(" input)
+                                        (str:ends-with-p ")" input))
+                                   input
+                                   (str:concat "(" input ")")))
+                 (expression (read-from-string proper-input))
```
Comment (Ambrevar) [inline source/command-commands.lisp:67]: Use the safe version, lest the user accidentally breaks everything!

**Review by Ambrevar — COMMENTED** [review]
Looks good to me!

**Comment by Ambrevar** [comment]
Please add some tests!  (Did you like by checking the checkbox? :p)

File: source/manual.lisp (hunk 1 above)
Comment (aartaka) [reply source/manual.lisp]: Because it's duplicated in the hooks section.

File: source/command-commands.lisp (hunk 2 above)
Comment (aartaka) [reply source/command-commands.lisp]: Done in 93b971c6fb80aea6117ec86c8a713842bed8e2f7!

**Comment by aartaka** [comment]
> Please add some tests! (Did you like by checking the checkbox? :p)

What kind of tests? Prompt input? Or proper command reading?

The checkbox meaning is two-fold, it's either "I've added tests", or "This code needs no tests, so I added all the necessary (0) tests." :P

**Comment by aartaka** [comment]
Tests: done (was easier than I expected. And one of the set-url tests fails on my machine xD)

Merging?

**Comment by Ambrevar** [comment]
Will test on Monday!  Remind me if I forget :)

**Comment by Ambrevar** [comment]
Tested, nice!

- Name the source "Lisp expression" for instance, something more meaningful than "Extended command".
- Why automatically surround with parentheses?  I'd leave it out and only compute the source suggestions if the first character is a `(`, thus making the whole prompt more efficient.
- This is currently not very usable because of some prompter limitations, namely that on each key press the focus is reset to the first source and the selection as well.  Related issues: #1666, #2033.

**Comment by aartaka** [comment]
> Which set-url test fails?
> 
> In my case, it seems to fail at random. There seems to be a race condition... :(

Yes, seems like so—it is random for me too. Merging this then?

**Comment by aartaka** [comment]
> Tested, nice!
> 
> * Name the source "Lisp expression" for instance, something more meaningful than "Extended command".

Done in 8f24038c536e040279e2d440ab2067a564174b79.

> * Why automatically surround with parentheses?  I'd leave it out and only compute the source suggestions if the first character is a `(`, thus making the whole prompt more efficient.

Seen this in some Interlisp demos and in Logo. It's convenient, so I thought it to be a virtue wrapping those if necessary.

> * This is currently not very usable because of some prompter limitations, namely that on each key press the focus is reset to the first source and the selection as well.  Related issues: [prompt-buffer: do-what-I-mean selection position during a single update #1666](https://github.com/atlas-engineer/nyxt/issues/1666), [Intelligently sorted prompt-buffer sources #2033](https://github.com/atlas-engineer/nyxt/issues/2033).

Indeed. What could help there, I think, is a better command-source filter, but yes, preserving cursor position would be nice.

**Comment by Ambrevar** [comment]
OK to merge!

**Comment by aadcg** [comment]
> inspiration comes from Interlisp shell that allowed to use Lisp expressions without the outer layer of parentheses and was extrememly helpful in suggesting code pieces

Can you point a reference? I'm curious.

**Comment by aartaka** [comment]
> > inspiration comes from Interlisp shell that allowed to use Lisp expressions without the outer layer of parentheses and was extrememly helpful in suggesting code pieces
> 
> Can you point a reference? I'm curious.

Oh, it wasn't Interlisp, it was Symbolics Lisp Machine reincarnated in OpenGenera, I guess. Here's the video: https://www.youtube.com/watch?v=o4-YnLpLgtk

**Comment by aadcg** [comment]
I'm not able to reproduce the functionality advertised in the screenshots. What am I missing?

**Comment by aadcg** [comment]
Ok, I just need to go the next sources. I see.

**Comment by aartaka** [comment]
What if you scroll to the bottom of the `execute-command` prompt?

**Comment by aartaka** [comment]
Exactly xD

**Comment by aadcg** [comment]
It seems to be that the fact that we omit the parenthesis doesn't allow for the possibility to have both variables and functions listed. 

For instance, when inputting "ppcre:" it only lists functions, but there are obviously variables defined in that namespace.

Should I open an issue?

**Comment by aadcg** [comment]
The logs are also a bit extensive. Perhaps it would make sense to shut them up?

```
<WARN> [14:54:46] Multiple 1= symbols found: NIL
<WARN> [14:54:48] Multiple 1=[ symbols found: NIL
<WARN> [14:54:48] Multiple 1=[+ symbols found: NIL
<WARN> [14:54:49] Multiple 1=[+[ symbols found: NIL
<WARN> [14:54:50] Multiple 1+ symbols found: NIL
<INFO> [14:54:54] Loading "https://github.com/atlas-engineer/nyxt".
<INFO> [14:54:54] Finished loading "https://github.com/atlas-engineer/nyxt".
<WARN> [14:55:34] Multiple I symbols found: NIL
<WARN> [14:55:34] Multiple IT symbols found: NIL
<WARN> [14:55:35] Multiple Q symbols found: NIL
<WARN> [14:55:35] Multiple QU symbols found: NIL
<WARN> [14:55:35] Multiple QUI symbols found: NIL
<WARN> [14:55:50] Multiple + symbols found: NIL
<INFO> [14:56:15] 6
<INFO> [14:57:24] Loading #P"/home/aadcg/.local/share/nyxt/bookmarks.lisp".
<WARN> [14:57:39] Multiple 2= symbols found: NIL
<WARN> [14:57:40] Multiple 2=[ symbols found: NIL
<WARN> [14:57:40] Multiple 2=[[ symbols found: NIL
<WARN> [14:57:40] Multiple 2=[[[ symbols found: NIL
<WARN> [14:58:00] Multiple N symbols found: NIL
<WARN> [14:58:01] Multiple NN symbols found: NIL
<WARN> [14:58:26] Multiple P symbols found: NIL
<WARN> [14:58:27] Multiple PP symbols found: NIL
<WARN> [14:58:31] Multiple PP symbols found: NIL
<WARN> [14:59:31] Multiple PPC symbols found: NIL
<WARN> [14:59:31] Multiple PPCR symbols found: NIL
<WARN> [14:59:31] Multiple PPCRE symbols found: NIL
<WARN> [14:59:33] Multiple PPCREN symbols found: NIL
<WARN> [14:59:33] Multiple PPCRENN symbols found: NIL
<WARN> [14:59:39] Multiple PPCRE symbols found: NIL
<WARN> [15:00:43] Multiple C symbols found: NIL
<WARN> [15:00:43] Multiple CL symbols found: NIL
<WARN> [15:00:44] Multiple CL- symbols found: NIL
<WARN> [15:00:44] Multiple CL-P symbols found: NIL
<WARN> [15:00:44] Multiple CL-PP symbols found: NIL
<WARN> [15:00:45] Multiple CL-PPC symbols found: NIL
<WARN> [15:00:45] Multiple CL-PPCR symbols found: NIL
<WARN> [15:00:45] Multiple CL-PPCRE symbols found: NIL
<WARN> [15:01:15] Multiple V symbols found: NIL
<WARN> [15:01:15] Multiple VE symbols found: NIL
<WARN> [15:01:15] Multiple VER symbols found: NIL
<WARN> [15:01:15] Multiple VERS symbols found: NIL
<WARN> [15:01:15] Multiple VERSI symbols found: NIL
<WARN> [15:01:15] Multiple VERSII symbols found: NIL
<WARN> [15:01:16] Multiple VERSIIO symbols found: NIL
<WARN> [15:01:17] Multiple VERSIO symbols found: NIL
<WARN> [15:01:17] Multiple VERSION symbols found: NIL
```

**Comment by aartaka** [comment]
Maybe yes. It's `resolve-symbol` being overly verbose (^_^;)

**Pull request merged**